File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        i, j, ans = 0,0,0
        n = len(s)
        characters = {'a':0,'b':0,'c':0}
        while j < n:
            characters[s[j]]+=1
            if all(characters.values()):
                ans += n-j
                characters[s[i]]-=1
                i+=1
            while(all(characters.values()) and i < n):
                characters[s[i]]-=1
                ans += n-j
                i+=1
            j+=1

        return ans

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n=s.size(),res=0,i=0;
        map<char,int>mpp;
        for(int j=0;j<n;j++){
            mpp[s[j]]++;
            if(mpp.size()==3){
                res+=n-j;
                mpp[s[i]]--;
                if(mpp[s[i]]==0)
                    mpp.erase(s[i]);
                i++;
            }
            while(mpp.size()==3 && i<n){
                mpp[s[i]]--;
                if(mpp[s[i]]==0)
                    mpp.erase(s[i]);
                res+=n-j;
                i++;
            }
        }
        return res;
    }
};